### PR TITLE
feat(solver): c++ letter agent

### PR DESCRIPTION
**Pull request by MksmOrlov**
Author: @CAT-i0n 

To test:
Comment _letter_message_processing_program.gwf_ and uncomment _message_processing_program.gwf_ in repo.path.
Use YTCSABRVXWY7DI3V7NGAFR2EUDRXK3DV wit token.
Type "Что в примере начинается на букву А?" or use any other letter.

### CODE REVIEW

File: docs/agents/letterAgent.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+LetterAgent is an agent added as example to show how to add new agents to system.
```
Comment (MksmOrlov) [inline docs/agents/letterAgent.md:1]:
```suggestion
LetterAgent is an agent added as an example to show how to add new agents to the system. 
```

File: docs/agents/images/letterSearchExample.png
Comment (MksmOrlov) [inline docs/agents/images/letterSearchExample.png:1]: Do we need a lang class?

File: docs/agents/letterAgent.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+<img src="../images/letterSearchExample.png"></img>
+
+To test you can use message like `Что в примере начинается на букву А?` or use any other letter.
+
+### Result
+
+Possible result codes:
+
+* `SC_RESULT_OK` - agent successfully found elements and generate phrase;
+* `SC_RESULT_ERROR`- `entity` is not found.
```
Comment (MksmOrlov) [inline docs/agents/letterAgent.md:35]:
```suggestion
* `SC_RESULT_ERROR`- `message node` or `entity` is not found.
```

File: docs/agents/letterAgent.md (hunk 1 above)
Comment (MksmOrlov) [inline docs/agents/letterAgent.md:1]: Write something about how we can use this agent (cities game and something else)

File: docs/agents/images/concept_phrase_about_letter.png
[the remark this answers is not in the document]
Comment (CAT-i0n) [reply docs/agents/images/concept_phrase_about_letter.png:1]: Not sure. Do we need to emphasize this in the images?

File: docs/agents/images/letterSearchExample.png
Comment (CAT-i0n) [reply docs/agents/images/letterSearchExample.png:1]: added it

File: docs/agents/letterAgent.md [hunk 3]
```diff
@@ -32,4 +33,8 @@ To test you can use message like `Что в примере начинается
 Possible result codes:
 
 * `SC_RESULT_OK` - agent successfully found elements and generate phrase;
-* `SC_RESULT_ERROR`- `entity` is not found.
\ No newline at end of file
+* `SC_RESULT_ERROR`- `message node` or `entity` is not found.
+
+### Usage
+
+The agent can be used as a template to write other agents, for example to cities game where we need to search words by last letter of previous word or for other suitable purposes.
```
Comment (MksmOrlov) [inline docs/agents/letterAgent.md:40]: Move usage section to the first sentence of the page

**Review by MksmOrlov — COMMENTED** [review]
check unresolved

**Review by MikhailSadovsky — APPROVED** [review]

**Review by MksmOrlov — COMMENTED** [review]
check the last unresolved

File: problem-solver/cxx/exampleModule/agent/LetterAgent.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  return SC_RESULT_OK;
+}
+
+bool LetterAgent::checkActionClass(ScAddr const & actionAddr)
+{
+  return m_memoryCtx.HelperCheckEdge(LetterKeynodes::action_letter_search, actionAddr, ScType::EdgeAccessConstPosPerm);
+}
+
+ScAddrVector LetterAgent::createAnswer(std::string const & linkContent)
+    const
```
Comment (MksmOrlov) [inline problem-solver/cxx/exampleModule/agent/LetterAgent.cpp:131]:
```suggestion
ScAddrVector LetterAgent::createAnswer(std::string const & linkContent) const
```

File: problem-solver/cxx/exampleModule/agent/LetterAgent.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+{
+  ScAddr const & answerLink = m_memoryCtx.CreateLink(ScType::LinkConst);
+  m_memoryCtx.SetLinkContent(answerLink, linkContent);
+  ScAddr const & edgeAccessConstPosPerm = m_memoryCtx.CreateEdge(ScType::EdgeAccessConstPosPerm, LetterKeynodes::word_starts_with_required_letter_answer_phrase, answerLink);
+  return {answerLink, edgeAccessConstPosPerm};
```
Comment (MksmOrlov) [inline problem-solver/cxx/exampleModule/agent/LetterAgent.cpp:136]:
```suggestion
  return {answerLink, edgeAccessConstPosPerm, LetterKeynodes::word_starts_with_required_letter_answer_phrase};
```

File: problem-solver/cxx/exampleModule/agent/LetterAgent.hpp [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+{
+  // Тип события условия инициирования sc-агента -- появление в sc-памяти выходящей дуги из `question_initiated` к действию
+  SC_CLASS(Agent, Event(scAgentsCommon::CoreKeynodes::question_initiated, ScEvent::Type::AddOutputEdge))
+
+  SC_GENERATED_BODY();
+
+private:
+  bool checkActionClass(ScAddr const & actionAddr);
+
+	ScAddrVector createAnswer(std::string const & linkContent) const;
```
Comment (MksmOrlov) [inline problem-solver/cxx/exampleModule/agent/LetterAgent.hpp:23]: fix formatting

File: problem-solver/cxx/exampleModule/keynodes/LetterKeynodes.cpp [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#include "LetterKeynodes.hpp"
+
+namespace exampleModule
+{
+    ScAddr LetterKeynodes::action_letter_search;
+    ScAddr LetterKeynodes::concept_message_about_letter_search;
+    ScAddr LetterKeynodes::rrel_entity;
```
Comment (MksmOrlov) [inline problem-solver/cxx/exampleModule/keynodes/LetterKeynodes.cpp:7]: Do not duplicate keynodes. Use `MessageKeynodes::rrel_entity`

Comment (MksmOrlov) [inline problem-solver/cxx/exampleModule/keynodes/LetterKeynodes.cpp:7]: fix formattinmg. Use on;y one tab (2 spaces width)

File: repo.path [hunk 8]
```diff
@@ -1,9 +1,9 @@
 # Change start page
 !kb/ims.ostis.kb/ui/ui_start_sc_element.scs
 !kb/ims.ostis.kb/ims/knowledge_base_IMS.scs
-
-# Do not build examples
```
Comment (MksmOrlov) [inline repo.path:5]: why do you remove my comment?

File: repo.path [hunk 9]
```diff
@@ -1,9 +1,9 @@
...
 !kb/pattern
+#!kb/section_subject_domain_of_dialogues/section_subject_domain_of_dialog_control/examples/message_processing_program.gwf
+!kb/section_subject_domain_of_dialogues/section_subject_domain_of_dialog_control/examples/letter_message_processing_program.gwf
```
Comment (MksmOrlov) [inline repo.path:6]: I think we shouldn't change program

File: kb/extra/letter_search/example.scs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+Aword
+=> nrel_main_idtf:
+    [Апельсин]
+        (* <- lang_ru;; *);;
+
+Mword
+=> nrel_main_idtf:
+    [Мандарин]
+        (* <- lang_ru;; *);;
```
Comment (MksmOrlov) [inline kb/extra/letter_search/example.scs:25]: Add newline please

**Review by MksmOrlov — COMMENTED** [review]
image
Do not rech the 120 chars line. You can format code automatically with **./scripts/clang/format_code.sh**

**Review by MksmOrlov — COMMENTED** [review]
Approve

**Review by CAT-i0n — APPROVED** [review]

**Pull request merged**